### ISSUE TITLE
LinkSyntax and ImageSyntax Incorrectly Inserted Into InlineSyntax Stack

### ISSUE THREAD

**Issue by mjordan56** [dart-lang/markdown#302]
The InlineParser constructor in inline_parser.dart initializes the TagState stack and sets up the list of inline syntax types to be invoked during the parsing process. The list of inline syntax objects first adds the inline syntaxes from the Document, then a generic text syntax, the default inline syntaxes, and finally the HTML syntaxes if the Document has enabled the HTML encoded flag. The final insertion into the inline syntax list is the insertion of the LinkSyntax and ImageSyntax objects at index position 1 and 2, respectively. Here is the code from inline_parser.dart:

```
InlineParser(this.source, this.document) : _stack = <TagState>[] {
    // User specified syntaxes are the first syntaxes to be evaluated.
    syntaxes.addAll(document.inlineSyntaxes);

    var hasCustomInlineSyntaxes = document.inlineSyntaxes
        .any((s) => !document.extensionSet.inlineSyntaxes.contains(s));

    // This first RegExp matches plain text to accelerate parsing. It's written
    // so that it does not match any prefix of any following syntaxes. Most
    // Markdown is plain text, so it's faster to match one RegExp per 'word'
    // rather than fail to match all the following RegExps at each non-syntax
    // character position.
    if (hasCustomInlineSyntaxes) {
      // We should be less aggressive in blowing past "words".
      syntaxes.add(TextSyntax(r'[A-Za-z0-9]+(?=\s)'));
    } else {
      syntaxes.add(TextSyntax(r'[ \tA-Za-z0-9]*[A-Za-z0-9](?=\s)'));
    }

    syntaxes.addAll(_defaultSyntaxes);

    if (document.encodeHtml) {
      syntaxes.addAll(_htmlSyntaxes);
    }

    // Custom link resolvers go after the generic text syntax.
    syntaxes.insertAll(1, [
      LinkSyntax(linkResolver: document.linkResolver),
      ImageSyntax(linkResolver: document.imageLinkResolver)
    ]);
  }
```

The insertion of LinkSyntax and ImageSyntax into the inline syntaxes list at position 1 seems wrong. Reading the comments, it seems like the intent was to insert these two inline syntax objects immediately after generic text syntax like this:

```
InlineParser(this.source, this.document) : _stack = <TagState>[] {
    // User specified syntaxes are the first syntaxes to be evaluated.
    syntaxes.addAll(document.inlineSyntaxes);

    var hasCustomInlineSyntaxes = document.inlineSyntaxes
        .any((s) => !document.extensionSet.inlineSyntaxes.contains(s));

    // This first RegExp matches plain text to accelerate parsing. It's written
    // so that it does not match any prefix of any following syntaxes. Most
    // Markdown is plain text, so it's faster to match one RegExp per 'word'
    // rather than fail to match all the following RegExps at each non-syntax
    // character position.
    if (hasCustomInlineSyntaxes) {
      // We should be less aggressive in blowing past "words".
      syntaxes.add(TextSyntax(r'[A-Za-z0-9]+(?=\s)'));
    } else {
      syntaxes.add(TextSyntax(r'[ \tA-Za-z0-9]*[A-Za-z0-9](?=\s)'));
    }

    // Custom link resolvers go after the generic text syntax.
    syntaxes.addAll([
      LinkSyntax(linkResolver: document.linkResolver),
      ImageSyntax(linkResolver: document.imageLinkResolver)
    ]);

    syntaxes.addAll(_defaultSyntaxes);

    if (document.encodeHtml) {
      syntaxes.addAll(_htmlSyntaxes);
    }
  }
```
I believe the original intent was to keep the document specified custom and extension set inline syntax objects at the top of the inline parser's processing list. By inserting LinkSyntax and ImageSyntax at index 1 in the list injects these two inline syntax parsers into the document supplied inline syntax objects which give them precedence over the document supplied inline syntax parsers.

I found this issue while debugging a problem with the flutter_markdown package. Depending on where certain inline syntax parsers were supplied with the document different output would be generated because the LinkSyntax object was consuming Markdown tags meant for a custom inline syntax parser. As such, the custom inline syntax parser wasn't being called as expected.

**Comment by srawlins** [comment]
Huh, your logic is certainly sound. Can you provide some markdown text which this bug causes to render incorrectly?

**Comment by mjordan56** [comment]
@srawlins Hi Sam. It's not the Markdown text as much as how the Markdown document parser is being configured to parse the Markdown document. The issue I found was basically this case:

1. I have a Markdown document that contains task list tags. For example, somewhere in my document I have:
```
- [ ] Task list item 1
- [ ] Task list item 2
- [ ] Task list item 3
```

2. I'm using the flutter_markdown package and specify a Markdown widget in my Flutter app like so:
```
      child: Markdown(
        data: markdownData,
        extensionSet: ExtensionSet.commonMark,
      ),
```
The output text does not parse the task list tags as they are not supported in the commonMark extension set.

3. I decide to create an inline syntax parser to handle task list tags (forget for a moment that this should be a block syntax parser. It doesn't matter for this example and is the real-world problem I found in the flutter_markdown package.) So I create an extension set with my TaskListSyntax type and include the commonMark parsers like this:
```
    var myExtensionSet = ExtensionSet(
      [...ExtensionSet.commonMark.blockSyntaxes],
      [TaskListSyntax(), ...ExtensionSet.commonMark.inlineSyntaxes],

   ...

      child: Markdown(
        data: markdownData,
        extensionSet: myExtensionSet,
      ),
```
Everything works! Woo-hoo, I have task list items in my Markdown widget.

4. Now I decide to add emoji tags to my Markdown. With my newfound knowledge and confidence working with syntax parsers, I add the EmojiSyntax to my extension set:
```
    var myExtensionSet = ExtensionSet(
      [...ExtensionSet.commonMark.blockSyntaxes],
      [EmojiSyntax(), TaskListSyntax(), ...ExtensionSet.commonMark.inlineSyntaxes],

   ...

      child: Markdown(
        data: markdownData,
        extensionSet: myExtensionSet,
      ),
```
My task list no longer works! I get the tags as plain text in my output!?!

The reason why this fails is that TaskListSyntax is now item 1 in the inlineSyntaxes list supplied with the Document. When the InlineParser constructor built the TagState stack the LinkSyntax and ImageSyntax objects got injected into the stack and push TaskListSyntax from position 1 to position 3. Since our Markdown task list tag uses '[' for its pattern as well as LinkSyntax and ImageSyntax, LinkSyntax wins, consumes the tag, and the TaskListSyntax parser never gets called. LinkSyntax sees the task list tags as a poorly formed link tag and outputs it as plain text.

There are a number of other ways to run into this problem, this is just the one I found but you can easily see how being position dependent in the inlineSyntaxes list can be problematic. As a test, I implemented my proposed solution of injecting LinkSyntax and ImageSyntax after the generic text syntax parser and the problem goes away because all of the user specified custom parsers are called first. Caveat emptor, I didn't do extensive (no pun intended) testing.

I'm still scratching my head as to the reasoning for adding LinkSyntax and ImageSyntax at the end of the constructor and setting their insertion point to index = 1. I'm sure someone had a reason at some point but it's lost to me.

**Comment by srawlins** [comment]
Ah, I see. Thanks for the explanation. Seems valid.

(It's too bad that [flutter markdown](https://pub.dev/documentation/flutter_markdown/latest/flutter_markdown/Markdown-class.html) doesn't let you pass individual syntaxes. You have to create your own extension set. The API that the markdown package provides is to pass block syntaxes and inline syntaxes to [Document](https://pub.dev/documentation/markdown/latest/markdown/Document-class.html), but the end result is the same order of things that you've shown above.)

Looks like this code was added ~7 years ago, when the plain word TextSyntax _was_ the first syntax in the list. https://github.com/dart-lang/markdown/blame/b8e319add581e5500ecde505705ae3ef52650840/lib/src/inline_parser.dart. So it seems pretty clear that that positioning is no longer correct.

**Comment by mjordan56** [comment]
@srawlins  Yeah, I agree that flutter_markdown should include parameters to Markdown and MarkdownBody for supplying blockSyntaxes and inlineSyntaxes. I submitted issue #259 with that exact request ([MarkdownWidget and Its Descendants Need to Expose blockSyntaxes and inlineSyntaxes](https://github.com/flutter/flutter_markdown/issues/259)).

So, since we agree the current index = 1 injection of LinkSyntax and ImageSyntax is probably incorrect, would you like me to submit a PR with the change? I think it would be great if this could be included in your 3.0 release.

**Comment by srawlins** [comment]
Certainly. Thanks much!
